### ISSUE TITLE
250819 Pytorch 이윤서

### ISSUE THREAD

**Issue by riicoseo** [AIBootcamp14/wiki-rider#33]
## 📅 학습 날짜  
*2025-08-19*

## 📌 키워드 및 요약 정리
- 텐서(Tensor) : 데이터를 담는 다차원 배열로, 딥러닝, Pytorch 에서 모든 데이터를 표현하는 기본 단위
- DNN(Deep Neural Network) : 여러 층으로 쌓인 신경망으로, 복잡한 패턴을 학습 가능
- CNN(Convolutional Neural Network): 이미지 처리에 특화된 신경망으로, 필터를 사용해 특징을 추출
- RNN(Recurrent Neural Network) : 순서가 있는 데이터(시계열, 텍스트)를 처리하는 신경망
- 역전파(Backpropagation) : 출력에서 입력 방향으로 오차를 전달해 가중치를 업데이트하는 과정
- 합성곱(Convolution): 필터가 이미지를 슬라이딩하면서 특징을 추출하는 연산
- 순환 구조(Recurrent) : 이전 시점의 정보를 현재 시점에 전달하는 구조

### 1. 텐서(Tensor) 

텐서는 딥러닝의 기본 데이터 구조로, 스칼라(0차원), 벡터(1차원), 행렬(2차원), 그리고 더 높은 차원의 배열을 통합한 개념

- PyTorch 코드 예시:
```python
import torch

scalar = torch.tensor(5.0)          # 0차원 (스칼라)
vector = torch.tensor([1, 2, 3])    # 1차원 (벡터)
matrix = torch.tensor([[1, 2], [3, 4]])  # 2차원 (행렬)
tensor_3d = torch.randn(2, 3, 4)    # 3차원 텐서

print(f"스칼라 shape: {scalar.shape}")     # torch.Size([])
print(f"벡터 shape: {vector.shape}")       # torch.Size([3])
print(f"행렬 shape: {matrix.shape}")       # torch.Size([2, 2])
print(f"3차원 텐서 shape: {tensor_3d.shape}")  # torch.Size([2, 3, 4])
```

- 주요 텐서 연산들:
reshape/view: 텐서의 모양 변경
squeeze/unsqueeze: 차원 추가/제거
concatenate/stack: 텐서 결합
transpose: 차원 순서 바꾸기
  
```python
# 텐서 모양 변경 예시
x = torch.randn(12)  # 1차원 텐서 (12개 원소)
x_matrix = x.view(3, 4)  # 3x4 행렬로 변경
x_3d = x.view(2, 2, 3)   # 2x2x3 3차원 텐서로 변경

# 차원 조작
x_expanded = x.unsqueeze(0)  # 차원 추가: (12,) → (1, 12)
x_squeezed = x_expanded.squeeze(0)  # 차원 제거: (1, 12) → (12,)
```

### 2. DNN(Deep Neural Network) 

DNN은 입력층, 여러 개의 은닉층, 출력층으로 구성된 다층 신경망.
각 층은 완전 연결(fully connected)되어 있어 이전 층의 모든 뉴런이 다음 층의 모든 뉴런과 연결.

- DNN 핵심 구성 요소
선형 변환: y= Wx + b    (가중치 행렬 곱셈 + 편향 더하기)
활성화 함수: ReLU, Sigmoid, Tanh 등으로 비선형성 추가
역전파: 오차를 역방향으로 전파하여 가중치 업데이트

```python
import torch.nn as nn

class SimpleDNN(nn.Module):
    def __init__(self, input_size, hidden_size, output_size):
        super(SimpleDNN, self).__init__()
        self.fc1 = nn.Linear(input_size, hidden_size)   # 첫 번째 은닉층
        self.fc2 = nn.Linear(hidden_size, hidden_size)  # 두 번째 은닉층
        self.fc3 = nn.Linear(hidden_size, output_size)  # 출력층
        self.relu = nn.ReLU()  # 활성화 함수
        
    def forward(self, x):
        x = self.relu(self.fc1(x))  # 첫 번째 층 + ReLU
        x = self.relu(self.fc2(x))  # 두 번째 층 + ReLU
        x = self.fc3(x)             # 출력층 (활성화 함수 없음)
        return x

# 모델 생성 및 사용 예시
model = SimpleDNN(input_size=784, hidden_size=128, output_size=10)  # MNIST 분류용
input_data = torch.randn(32, 784)  # 배치 크기 32, 입력 차원 784
output = model(input_data)  # 순전파
print(output.shape)  # torch.Size([32, 10])
```

### 3. CNN(Convolutional Neural Network) - 합성곱 신경망

CNN은 이미지의 공간적 구조를 유지하면서 특징을 추출하는 신경망.
합성곱층, 풀링층, 완전연결층으로 구성됨.

- DNN 핵심 구성 요소
1) 합성곱층(Convolutional Layer) : 필터(커널)가 입력 이미지 위를 슬라이딩하면서 특징 추출
2) 풀링층(Pooling Layer) : 특징맵의 크기를 줄여 계산량 감소. 위치 불변성 제공.

```python
class SimpleCNN(nn.Module):
    def __init__(self):
        super(SimpleCNN, self).__init__()
        # 특징 추출 부분
        self.conv1 = nn.Conv2d(3, 32, 3, padding=1)
        self.conv2 = nn.Conv2d(32, 64, 3, padding=1)
        self.pool = nn.MaxPool2d(2, 2)
        
        # 분류 부분
        self.fc1 = nn.Linear(64 * 8 * 8, 128)
        self.fc2 = nn.Linear(128, 10)
        self.relu = nn.ReLU()
        
    def forward(self, x):
        # 특징 추출
        x = self.pool(self.relu(self.conv1(x)))  # 32x32 → 16x16
        x = self.pool(self.relu(self.conv2(x)))  # 16x16 → 8x8
        
        # 평탄화 및 분류
        x = x.view(-1, 64 * 8 * 8)  # 1차원으로 펼치기
        x = self.relu(self.fc1(x))
        x = self.fc2(x)
        return x
```

### 4. RNN(Recurrent Neural Network) - 순환 신경망

RNN은 순서가 있는 데이터(시계열, 자연어)를 처리하기 위해 설계된 신경망.
이전 시점의 은닉 상태를 현재 시점에 전달하는 순환 구조.

```python
# 기본 RNN 예시
rnn_layer = nn.RNN(input_size=10,    # 입력 차원
                   hidden_size=20,   # 은닉 상태 차원
                   num_layers=2,     # RNN 층 수
                   batch_first=True) # 배치를 첫 번째 차원으로

# 입력: (배치_크기, 시퀀스_길이, 입력_차원)
input_seq = torch.randn(32, 15, 10)  # 32개 샘플, 길이 15, 차원 10
output, hidden = rnn_layer(input_seq)
print(f"출력 shape: {output.shape}")      # torch.Size([32, 15, 20])
print(f"은닉상태 shape: {hidden.shape}")   # torch.Size([2, 32, 20])
```

- RNN 한계 & 해결
1) 기울기 소실 문제: 긴 시퀀스에서 과거 정보가 사라짐
2) 해결책: LSTM, GRU 같은 개선된 구조 사용

### 5. 역전파와 경사하강법
모든 신경망의 학습은 손실함수의 기울기를 계산하여 가중치를 업데이트하는 과정

- 수식:

<img width="373" height="148" alt="Image" src="https://github.com/user-attachments/assets/fe870950-e6db-4d1e-8a9a-06913956ad7b" />

η: 학습률(learning rate)
L: 손실함수
∂L/∂W​: 가중치에 대한 손실함수의 기울기

## ❓ 어려웠던 부분  
- 텐서 차원 변환: 처음 보는 reshape, view, transpose 등의 차원 조작이 헷갈림
- 역전파의 연쇄법칙: 복잡한 신경망에서 기울기가 어떻게 전파되는지 수식적으로 이해하기 어려웠음.
- 이론과 별개로 pytorch 로 코드 구현하는 부분이 어려웠음.

## 💬 질문 또는 토론 유도  
- CNN vs RNN: 시계열 이미지 데이터(비디오)를 처리할 때 CNN과 RNN 중 어떤 것이 더 적합할까요? 아니면 둘을 결합한 구조가 필요할까요?

## 🔗 참고 자료 
- [Pytorch  공식문](https://pytorch.org/docs/stable/index.html)
- [Pytorch 튜토리얼](https://docs.pytorch.org/tutorials/)

**Comment by parkjunohjjang12** [comment]
안녕하세요, 이윤서 님! 올려주신 내용 잘 읽었습니다.
텐서, DNN, CNN, RNN, 그리고 backpropagation까지 딥러니의 개념들을 PyTorch 코드 예시와 함께 정리해주셔서 이해하기 쉽게 잘 정리된 것 같습니다. 특히 코드와 설명을 병행해준 부분이 실습과 연결하기에 좋아 보입니다.

---
## 1️⃣ 어려웠던 부분 피드백

**텐서 차원 변호나(reshpae, view, transpose 등)**

- 헷갈리기 쉽지만 자주 쓰이는 연산입니다. 앞서 박재홍 님께서도 비슷한 질문을 하셔서 답변을 드렸는데, view는 메모리가 연속해야 한다는 제약이 있고, reshape는 더유연하게 동작한다고 보시면 될 것 같습니다. transpose와 permute는 차원 순서를 바꾸는 데 쓰이고, 3차원 이상의 고차원의 경우 permute를 주로 더 사용한다고 생각하시면 도리 것 같습니다.

**backpropagation과 연쇄법칙**
- 복잡한 신경망에서 기울기 전파가 어떻게 이루어지는지 수식적으로 이해하기 어렵다는 부분은 자연스러운 어려움인 것 같습니다. PyTorch의 autograd가 이 과정을 자동으로 해주지만, 직접 2~3층 MLP를 만들어 손으로 미분 과정을 풀어보거나, scratch부터 구현을 해 보는 것을 추천 드립니다!
- 또한, 이에 대한 강의도 하나 공유 드립니다. 이전에 민변호님과 정서우님께도 공유드린 자료인데, backpropagation에 대한 자세한 내용을 다루니, 한 번 들어보시는 것을 추천합니다!
📄  [이준석 교수님 ML/DL Lec3: Neural Networks & Backpropagation](https://www.youtube.com/watch?v=O8FMi94Upxc&list=PL0E_1UqNACXAc-tb9z2069eDeIOpcO8A2&index=3
)
📄 [CS231n](https://cs231n.github.io/optimization-2/)  
📄 [Backpropagation Calculus – 3Blue1Brown](https://www.3blue1brown.com/lessons/backpropagation-calculus)  
- 또한 아래 자료는 딥러닝에 대한 기초 내용을 다룹니다. 뿐만 아니라, backpropagation 코드 구현도 포함되어있으니 꼭 직접 구현해보시는 것을 추천합니다!
📄 [NYU Deep Learning](https://atcold.github.io/NYU-DLSP21/)

**PyTorch 코드 구현**
- 처음에는 문법적 요소가 낯설 수 있습니다. 이론적으로 이해가 되더라도 실제로 구현하는 것은 또 다른 것이라고 생각합니다. 하지만 제 생각에는 이러한 방법에는 지름길은 없는 것 같습니다. MNIST와 같은 작은 이미지 분류모델부터 구현을 시작해보고, 그 후에 kaggle과 같은 competition을 자주 해보시거나, 논문 구현 등을 해 보면 어느순간 코드 구현에 충분히 익숙해지실 수 있을거라 생각합니다!

---
**CNN vs RNN: 시계열 이미지 데이터(비디오)를 처리할 때 어떤 것이 적합할까요?**
- 결론부터 이야기 드리자면, 최근 몇 년간은 CNN, RNN이 아닌 Transformer가 가장 많이 사용됩니다. Transformer 자체가 결국은 Attention 매커니즘을 통해 시공간의 모든 위치 간 관계를 직접 학습할 수 있어서, 비디오의 복잡한 시간적 의존성을 더 효과적으로 포착하기 때문입니다.
- 만약 CNN과 RNN 둘 중에 하나를 선택한다면 CNN은 프레임 단위의 공간적(단일 이미지)의 특징을 추출하는데 강점이 있고, RNN은 시간적(현재 프레임 → 다음 프레임)에 대해서 모델링하는 데 적합합니다. 따라서 제 생각에는 짧은 비디오의 경우 CNN이, 긴 영상의 비디오의 경우 RNN이 더 적합할 것으로 보입니다.
- 비디오 처리의 전통적 접근은 CNN + RNN을 결합하는 것입니다. CNN으로 공간 정보를 추출하고, RNN으로 시간적 변화를 모델링하는 구조죠.
- 다만 앞서 결론을 말씀 드렸듯이, 최근에는 Video Transformer 가 가장 많이 사용되니 이에 대해서 공부하는 것이 좋을 것으로 보입니다!
📄  [Video Swin Transformer](https://arxiv.org/abs/2106.13230)
📄  [ViViT: A Video Vision Transformer](https://arxiv.org/abs/2103.15691)
---
지금처럼 코드와 개념을 같이 정리하는 습관은 정말 좋습니다. 이에 대해서 꾸준히 하시면 더 많이 발전하실 수 있을거라고 생각합니다!  
화이팅입니다! 💪
